### PR TITLE
feat(jasmine_node_test): add ibazel support

### PR DESCRIPTION
**Pull request by Toxicable**

### CODE REVIEW

File: internal/jasmine_node_test/jasmine_runner.js [hunk 1]
```diff
@@ -11,45 +12,94 @@ const UTF8 = {
...
+  // Remove the manifest, some tested code may process the argv.
+  process.argv.splice(2, 1)[0];
+  
+  const watch = true;
+  const jrunner = createRunner(watch);
+  addFiles(manifestLocation, jrunner);
 
-  jrunner.execute();
-  return 0;
+  // TODO: find out how to determine if we're running under ibazel
```
Comment (alexeagle) [inline internal/jasmine_node_test/jasmine_runner.js]: use the environment variable
https://github.com/bazelbuild/bazel-watcher/pull/60/files#diff-79db8648ad38b76a407bbc7744e0a7aeR74

File: internal/jasmine_node_test/jasmine_node_test.bzl [hunk 2]
```diff
@@ -41,10 +41,12 @@ def jasmine_node_test(name, srcs = [], data = [], deps = [], **kwargs):
   all_data += [Label("//internal/jasmine_node_test:jasmine_runner.js")]
   all_data += [":%s_devmode_srcs.MF" % name]
   entry_point = "build_bazel_rules_nodejs/internal/jasmine_node_test/jasmine_runner.js"
+  tags = ["ibazel_notify_changes"]
 
   nodejs_test(
       name = name,
       data = all_data,
+      tags = tags,
```
Comment (alexeagle) [inline internal/jasmine_node_test/jasmine_node_test.bzl]: this prevents the user adding other tags on the target - should amend this macro to have argument `tags = []` and add the ibazel one to the given ones

File: internal/jasmine_node_test/jasmine_runner.js [hunk 3]
```diff
@@ -11,45 +12,113 @@ const UTF8 = {
...
+      console.error('ibazel failed to stop jasmine after 5s; probably a bug');
+      process.exit(1);
+    }, 5000);
+  });
+}
+
+function deleteFromCache(fileName) {
+  // jasmine uses `require(fileName)` to read the files into the program
+  // so we have to delete them from the cache to ensure they read again on subsequent runs
+  // when in ibazel mode
```
Comment (alexeagle) [inline internal/jasmine_node_test/jasmine_runner.js]: be consistent whether it's called "watch mode" or "ibazel mode"

File: internal/jasmine_node_test/jasmine_runner.js [hunk 4]
```diff
@@ -11,45 +12,113 @@ const UTF8 = {
...
+  delete require.cache[require.resolve(fileName)];
+}
+
+function addFiles(manifestLocation, jrunner){
+  deleteFromCache(manifestLocation)
+
+  const manifest = require.resolve(manifestLocation);
+  const file = fs.readFileSync(manifest, UTF8);
```
Comment (alexeagle) [inline internal/jasmine_node_test/jasmine_runner.js]: seems like an extra read? what do you do with this variable?

File: internal/jasmine_node_test/jasmine_runner.js [hunk 5]
```diff
@@ -11,45 +12,113 @@ const UTF8 = {
...
+  fs.readFileSync(manifest, UTF8).split('\n').filter(l => l.length > 0).forEach(f => {
+    // ensure it's actually read from disk again
+    deleteFromCache(f);
+    try {
+      // try requiring it incase there are syntax errors that kills the program later on
+      require(f);
+      jrunner.addSpecFile(f);
+    } catch (error) {
+      // dont add the files to jasmine since it'll run them and it'll kill the process
+      console.error('Error reading file: ', require.resolve(f))
```
Comment (alexeagle) [inline internal/jasmine_node_test/jasmine_runner.js]: message should indicate that we are ignoring the file instead of running it as a spec

File: internal/jasmine_node_test/jasmine_runner.js [hunk 6]
```diff
@@ -11,45 +12,113 @@ const UTF8 = {
...
+  const file = fs.readFileSync(manifest, UTF8);
+  fs.readFileSync(manifest, UTF8).split('\n').filter(l => l.length > 0).forEach(f => {
+    // ensure it's actually read from disk again
+    deleteFromCache(f);
+    try {
+      // try requiring it incase there are syntax errors that kills the program later on
+      require(f);
+      jrunner.addSpecFile(f);
+    } catch (error) {
+      // dont add the files to jasmine since it'll run them and it'll kill the process
```
Comment (alexeagle) [inline internal/jasmine_node_test/jasmine_runner.js]: is this feature related to the watch mode? what happens with the code at HEAD when we run into some file that has syntax errors?
if it's orthogonal to the watch mode, let's move to different PR

File: internal/jasmine_node_test/jasmine_runner.js [hunk 7]
```diff
@@ -11,45 +12,113 @@ const UTF8 = {
...
+  const manifestLocation = args[0];
+  // Remove the manifest, some tested code may process the argv.
+  process.argv.splice(2, 1)[0];
+
+  // only watch when under ibazel
+  const watch = Boolean(process.env[IBAZEL_NOTIFY_CHANGES]);
 
-  jrunner.execute();
-  return 0;
+  const jrunner = createRunner(watch);
```
Comment (alexeagle) [inline internal/jasmine_node_test/jasmine_runner.js]: instead of passing a boolean here, how about passing an onComplete callback? that way anything to do with watch mode is localized here.
As you have it, there is stuff about exit codes in the onComplete above which is unused in single-run mode

Comment (Toxicable) [reply internal/jasmine_node_test/jasmine_runner.js]: I've moved all of the watch logic into `iBazelMode`

File: internal/jasmine_node_test/jasmine_runner.js (hunk 5 above)
Comment (Toxicable) [reply internal/jasmine_node_test/jasmine_runner.js]: done

File: internal/jasmine_node_test/jasmine_runner.js (hunk 4 above)
Comment (Toxicable) [reply internal/jasmine_node_test/jasmine_runner.js]: whops, was debugging with that, fixed

File: internal/jasmine_node_test/jasmine_runner.js (hunk 3 above)
Comment (Toxicable) [reply internal/jasmine_node_test/jasmine_runner.js]: changed all to iBazelMode

File: internal/jasmine_node_test/jasmine_node_test.bzl (hunk 2 above)
Comment (Toxicable) [reply internal/jasmine_node_test/jasmine_node_test.bzl]: fixed

File: internal/jasmine_node_test/jasmine_runner.js (hunk 6 above)
Comment (Toxicable) [reply internal/jasmine_node_test/jasmine_runner.js]: It's not directly related to watch mode, but you wouldn't see this specific error in any other scenario.
On HEAD the process only ever runes once, so it dosen't matter if jasmine tries using require since it fails the test anyway.

**Comment by Globegitter** [comment]
Is there anything still blocking this from being merged? Apart from the merge conflict?

**Comment by Toxicable** [comment]
@Globegitter I ran into a few issues with this not actually working correctly and had very little time recently to do any open source stuff.

**Comment by alexeagle** [comment]
@Toxicable need someone from our team to pick this up?

**Comment by Toxicable** [comment]
@alexeagle yeah that would be good on this one

**Comment by Globegitter** [comment]
@alexeagle I think this might be running into an issue with bazel/ibazel itself. See https://github.com/bazelbuild/bazel-watcher/issues/184 and https://github.com/bazelbuild/bazel/pull/8322

Would be good to get https://github.com/bazelbuild/bazel/pull/8322 pushed forward.

**Comment by Toxicable** [comment]
this probably needs to be rewritten, will redo if there's interest in the linked ticket

**Comment by joeljeske** [comment]
I am interested in this capability. I think it could greatly enhance the the DX while writing tests.

@Toxicable, sorry, what linked ticked are you referring to? (Both tickets mentioned by @Globegitter appear resolved at the time of this ticket closing?)

**Comment by Toxicable** [comment]
@joeljeske this would be the one https://github.com/bazelbuild/rules_nodejs/issues/177
